android/src/main/java/io/radar/react/RNRadarModule.java

```diff
@@ -42,59 +56,99 @@ public void setDescription(String description) {
+            }
+        } else {
+            promise.resolve(foreground ? "GRANTED_BACKGROUND" : "DENIED");
+        }
     }
 
     @ReactMethod
     public void requestPermissions(boolean background) {
         Activity activity = getCurrentActivity();
         if (activity != null) {
-            ActivityCompat.requestPermissions(activity, new String[] { Manifest.permission.ACCESS_FINE_LOCATION }, 0);
+            if (Build.VERSION.SDK_INT >= 23) {
+                int requestCode = 0;
+                if (Build.VERSION.SDK_INT >= 29) {
+                    ActivityCompat.requestPermissions(activity, new String[] { Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_BACKGROUND_LOCATION }, requestCode);
+                } else {
+                    ActivityCompat.requestPermissions(activity, new String[] { Manifest.permission.ACCESS_FINE_LOCATION }, requestCode);
+                }
+            }
         }
     }
 
     @ReactMethod
-    public void startTracking(ReadableMap optionsMap) {
-        Radar.startTracking(RNRadarUtils.optionsForMap(optionsMap));
-    }
+    public void getLocation(final Promise promise) {
+        Radar.getLocation(new Radar.RadarLocationCallback() {
+            @Override
+            public void onComplete(@NonNull Radar.RadarStatus status, @Nullable Location location, boolean stopped) {
+                if (promise == null) {
+                    return;
+                }
 
-    @ReactMethod
-    public void stopTracking() {
-        Radar.stopTracking();
+                if (status == Radar.RadarStatus.SUCCESS) {
+                    try {
+                        WritableMap map = Arguments.createMap();
+                        map.putString("status", status.toString());
+                        if (location != null) {
+                            map.putMap("location", RNRadarUtils.mapForJson(Radar.jsonForLocation(location)));
+                        }
+                        map.putBoolean("stopped", stopped);
+                        promise.resolve(map);
+                    } catch (JSONException e) {
+                        promise.reject(Radar.RadarStatus.ERROR_SERVER.toString(), Radar.RadarStatus.ERROR_SERVER.toString());
+                    }
+                } else {
+                    promise.reject(status.toString(), status.toString());
+                }
+            }
+        });
     }
 
     @ReactMethod
     public void trackOnce(final Promise promise) {
-        Radar.trackOnce(new RadarCallback() {
+        Radar.trackOnce(new Radar.RadarTrackCallback() {
             @Override
-            public void onComplete(@NonNull Radar.RadarStatus status, Location location, RadarEvent[] events, RadarUser user) {
-                if (promise == null)
+            public void onComplete(@NonNull Radar.RadarStatus status, @Nullable Location location, @Nullable RadarEvent[] events, @Nullable RadarUser user) {
+                if (promise == null) {
```

Comment:
wonder if this is the desirable default behavior if a `Promise` is not passed here - we can still execute the request / operation without resolving / rejecting back to the caller. I believe the Web SDK does it this way

Reply:
In this case, we are calling `trackOnce` even if promise is `null`. We're just skipping resolving/rejecting the promise. It's valid to call `trackOnce` to force a location update even if you're not handling the response.

For the other APIs, there's no point in making an API request if you're not going to handle the response, so we just short circuit. (See `getContext()`, for example.)